Article qdrant 1.2.x (#142)
qdrant/landing_page: qdrant-landing/content/articles/qdrant-1.2.x.md

don't want to make your payloads redundant by keeping them in separate collections.

![Optional vectors](/articles_data/qdrant-1.2.x/optional-vectors.png)

However, up to the previous version, we requested you provide all the vectors for each point. There 
have been many requests to allow nullable vectors, as sometimes you cannot generate an embedding or 
simply don't want to for reasons we don't need to know.

### Grouping requests

Embeddings are great for capturing the semantics of the documents, but we rarely encode larger pieces 
of data into a single vector. Having a summary of a book may sound attractive, but in reality, we 
divide it into paragraphs or some different parts to have higher granularity. That pays off when we 
perform the semantic search, as we can return the relevant pieces only. That's also how modern tools 
like Langchain process the data. The typical way is to encode some smaller parts of the document and 
keep the document id as a payload attribute.

![Query without grouping request](/articles_data/qdrant-1.2.x/without-grouping-request.png)

There are, however, some cases, in which we want to find the relevant parts, but no more than just 

> a single result per document. Up till now, we had to implement such a mechanism on the client side 

Reviewer: Not sure if we need to mention that we can generate group of different sizes, not only the top 1

Author: Good point, I'll mention that!

![Query with grouping request](/articles_data/qdrant-1.2.x/with-grouping-request.png)

### Nested filters

Unlike some other vector databases out there, Qdrant accepts any arbitrary JSON payload, including
arrays, objects, and arrays of objects. You can also [filter the search results using nested 
keys](/documentation/filtering/#nested-key), even though arrays (using the `[]` syntax). 

Before Qdrant 1.2 it was actually impossible to express some more complex conditions for the
nested structures. For example, let's assume we have the following payload:

```json
{
  "country": "Japan",
  "cities": [
    {
      "name": "Tokyo",
      "population": 9.3,
      "area": 2194
    },
    {
      "name": "Osaka",
      "population": 2.7,
      "area": 223
    },
    {
      "name": "Kyoto",
      "population": 1.5,
      "area": 827.8
    }
  ]
}
```

We want to filter out the results to include the countries with a city with over 2 million citizens 
and an area bigger than 500 square kilometers but no more than 1000. There is no such a city in 
Japan, looking at our data, but if we wrote the following filter, it would be returned:

```json
{
  "filter": {
    "must": [
      {

>         "key": "country.cities.population",

Reviewer: Have you tried this example?

I think you would need the `[]` notation here.

`"country.cities[].population"`

        "range": {
          "gte": 2
        }
      },
      {

>         "key": "country.cities.area",

Reviewer: same comment as above

                }
              },
              {
                "key": "area",
                "range": {
                  "gt": 500,
                  "lte": 1000
                }
              }
            ]
          }
        }
      }
    ]
  },
  "limit": 3
}
```

The syntax is consistent with all the other supported filters and enables new possibilities. In 
our case, it allows us to express the joined condition on a nested structure and make the results 
list empty but correct.

## Important changes

The latest release focuses not only on the new features but also introduces some changes making 
Qdrant even more reliable.

### Recovery mode

There has been an issue in memory-constrained environments, such as cloud, happening when users were 
pushing massive amounts of data into the service using `wait=false`. This data influx resulted in an 
overreaching of disk or RAM limits before the Write-Ahead Logging (WAL) was fully applied. This 
situation was causing Qdrant to attempt a restart and reapplication of WAL, failing recurrently due 
to the same memory constraints and pushing the service into a frustrating crash loop with many 
Out-of-Memory errors.

> Qdrant 1.2 enters the recovery mode, if enabled, when it detects the repeated failures on startup. 

Reviewer 2: ```suggestion
Qdrant 1.2 enters recovery mode, if enabled, when it detects a failure on startup. 
```

list empty but correct.

## Important changes

The latest release focuses not only on the new features but also introduces some changes making 
Qdrant even more reliable.

### Fixed memory errors on startup

There has been an issue in memory-constrained environments, such as cloud, happening when users were 
pushing massive amounts of data into the service using `wait=false`. This data influx resulted in an 
overreaching of disk or RAM limits before the Write-Ahead Logging (WAL) was fully applied. This 
situation was causing Qdrant to attempt a restart and reapplication of WAL, failing recurrently due 
to the same memory constraints and pushing the service into a frustrating crash loop with many 
Out-of-Memory errors.

Qdrant 1.2 automatically detects if memory usage exceeds a specific threshold, informs a user with an 
error message in the logs, and makes the service halt the loading of collection data and commence 
operations in a partial state. This state allows for removing collections but doesn't support search 
or update functions. If Qdrant enters this locked state, a reboot that either increases available or 

> reduces required resources resolves it.

Reviewer 2: It is not quite clear to me what this refers to.

I assume this bit refers to recovery mode. But I think this text doesn't really match the way it actually works:

- AFAIK there is no detection with a 'threshold', rather, if Qdrant crashes on startup due to memory errors it can restart in recovery mode if enabled
- recovery mode is disabled by default

More info here: https://github.com/qdrant/docs/blob/v1.2.x/qdrant/v1.2.x/administration.md#recovery-mode

Reviewer 2: Yeah, the original idea was to make this a lot more _automagic_. But detecting out-of-memory situations is extremely hard and error prone. Therefore we changed the approach and made it a lot simpler, as described above.

### Appendable mmap

For a long time, segments using mmap storage were `non-appendable` and could only be constructed by 
the optimizer. Dynamically adding vectors to the mmap file is fairly complicated and thus not 
implemented in Qdrant, but we did our best to implement it in the recent release. If you want 

> to read more about segments, check out our docs on [vector storage](/documentation/storage/#vector-storage).

Reviewer 2: Appendable mmap is indeed new in code, but if I am correct, there is no way for users to enable it at this moment.

Reviewer 2: Users should be able to use it. This is a code problem. Looking into it now.

Author: I thought it was controlled by the optimizer. @generall Do you think we should keep that info?

Reviewer 2: No I'm planning to work on that now.

I want to describe it in these two paragraphs:

- https://qdrant.tech/documentation/collections/#create-collection
- https://qdrant.tech/documentation/how_to/#bulk-upload-a-large-number-of-vectors

Reviewer 2: Opened a PR for it here: https://github.com/qdrant/docs/pull/122
